zanata-war/src/main/java/org/zanata/servlet/MultiFileUploadServlet.java

```diff
@@ -392,7 +392,23 @@ private DocumentFileUploadForm createUploadFormForItem(FileItem item) throws IOE
             form.setFirst(true);
             form.setLast(true);
             form.setSize(item.getSize());
-            form.setFileType(translationFileServiceImpl.extractExtension(item.getName()));
+
+            String extension = FileUtil.extractExtension(item.getName());
+            String fileType = null;
+            if(!fileTypes.isEmpty()) {
+                for(String parsedFileType: fileTypes) {
+                    DocumentType docType = DocumentType.valueOf(parsedFileType);
+                    if(docType != null && docType.getSourceExtensions().contains(extension)) {
+                        fileType = docType.name();
+                        break;
+                    }
+                }
+            }
+            fileType =
+                    fileType != null ? FileUtil
+                            .extractExtension(item.getName()) : fileType;
+
+            form.setFileType(fileType);
```

Comment:
- [ ] Should put all the new code just above here into a method, so that this line is something like:

```
form.setFileType(getFileTypeForItem(item));
```